protons/app/proposals.py

```diff
@@ -159,3 +160,296 @@ def propose_states(self, drive, residue_pool_indices):
+    """This class is a baseclass for selecting water molecules or ions to maintain charge neutrality."""
+
+    def propose_swaps(self, drive, net_charge_difference):
+        """Select a series of water molecules/ions to swap.
+
+        Parameters
+        ----------
+
+        drive - a ProtonDrive object with a swapper attached.
+        net_charge_difference - int, the total charge that needs to be countered by ions.
+
+        Nota bene
+        ---------
+        The net charge difference will have the opposite charge of the ions that will be added to the system,
+        such that the resulting total charge change will be 0.
+
+        Returns
+        -------
+        list(int) - indices of water molecule in saltswap that are swapped
+        list(tuple(int,int),) -  list of tuples with (initial, final) states of the water molecule/ion as int, one tuple for each requested swap
+            By saltswap convention
+             - 0 is water
+             - 1 is cation
+             - 2 is anion
+        float - log (probability of reverse proposal)/(probability of forward proposal)
+        """
+
+        return list(), list(), float()
+
+
+class UniformSwapProposal(_SaltSwapProposal):
+    """Uniformly selects as many water/ions as needed from the array of available waters/ions."""
+
+    def __init__(self, cation_coefficient: float = 0.5):
+        """Instantiate a UniformSwapProposal.
+
+        Parameters
+        ----------
+        cation_coefficient - float, optional. Must be between 0 and 1, default 0.5.
+            The fraction of the chemical potential of a water pair -> salt pair transformation that is attributed to the
+            cation -> water transformation.
+        """
+        if not 0.0 <= cation_coefficient <= 1.0:
+            raise ValueError("The cation coefficient should be between 0 and 1.")
+
+        self._cation_weight = cation_coefficient
+        self._anion_weight = 1.0 - cation_coefficient
+
+    @staticmethod
+    def _select_ion_water_swaps(drive, charge_to_counter):
+        """Select ions or water molecule swap procedure to facilitate maintaining charge neutrality while changing protonation states.
+
+        Notes
+        -----
+        Technically, for charge differences of 2 we could change one anion into a cation or vice versa.
+        This code would instead remove one anion, and change one water into a cation.
+
+        Parameters
+        ----------
+        drive - a ProtonDrive object
+        charge_to_counter - the total difference in charge that needs to be countered.
+
+        Returns
+        -------
+        dict(water_to_cation, water_to_anion, cation_to_water, anion_to_water)
+
+        """
+
+        # Note that we don't allow for direct transitions between ions of different charge.
+        swaps = dict(water_to_cation=0, water_to_anion=0, cation_to_water=0, anion_to_water=0)
+        excess_ions = int(drive.excess_ions)  # copy
+
+        while abs(charge_to_counter) > 0:
```

Comment:
Why the while loop rather than just figuring out what you need and drawing directly from that? I assume the while loop does one at a time?

Reply:
Oh, just looked down, yeah, it does. So is this just simpler to account for?

Reply:
This was conceptually simpler at the time of the original implementation. There might be a more clever way to do this now that I've simplified the calls to be functions. I'll take another look.